File: mobile/test/announcements_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:kaede_mobile/src/api/announcement_repository.dart';
import 'package:kaede_mobile/src/api/api_client.dart';
import 'package:kaede_mobile/src/api/kaede_repository.dart';
import 'package:kaede_mobile/src/auth/session_vault.dart';
import 'package:kaede_mobile/src/core/refs.dart';
import 'package:kaede_mobile/src/domain/announcements.dart';
import 'package:kaede_mobile/src/domain/models.dart';
import 'package:kaede_mobile/src/features/guild/announcement_management_tab.dart';
import 'package:kaede_mobile/src/protocol/generated.dart';

void main() {
  final actor = KaedeUser.fromJson(<String, Object?>{
    'id': '7',
    'origin_domain': 'chat.example',
    'username': 'maple',
    'handle': 'maple@chat.example',
  });

  test('filters cross-guild targets and enforces publish authority', () {
    final source = _channel(
      id: '2',
      type: 5,
      permissions: Permission.viewChannel |
          Permission.readMessageHistory |
          Permission.sendMessages,
    );
    final target = _channel(
      id: '3',
      type: 0,
      permissions: Permission.manageWebhooks,
    );
    final denied = _channel(id: '4', type: 0, permissions: 0);
    final e2eeRequired = _channel(
      id: '7',
      type: 0,
      permissions: Permission.manageWebhooks,
      e2eeRequired: true,
    );
    final remoteTarget = _channel(
      id: '5',
      type: 0,
      permissions: Permission.administrator,
      domain: 'remote.example',
      guildId: '9',
    );
    final remoteSource = _channel(
      id: '6',
      type: 5,
      permissions: Permission.sendMessages,
      domain: 'remote.example',
      guildId: '9',
    );
    final localGuild = _guild('1', 'Local', <KaedeChannel>[
      source,
      target,
      denied,
      e2eeRequired,
    ]);
    final remoteGuild = _guild(
      '9',
      'Remote',
      <KaedeChannel>[remoteTarget, remoteSource],
      domain: 'remote.example',
    );
    final post = KaedeMessage.fromJson(_messageJson());

    expect(canReadAnnouncementChannel(localGuild, source, actor), isTrue);
    expect(
      announcementTargets(<KaedeGuild>[localGuild, remoteGuild], actor)
          .map((item) => item.ref.wire),
      <String>['3@chat.example', '5@remote.example'],
    );
    expect(
      canPublishAnnouncementMessage(localGuild, source, post, actor),
      isTrue,
    );
    expect(
      canPublishAnnouncementMessage(
        localGuild,
        source,
        post.copyWith(flags: messageFlagCrossposted),
        actor,
      ),
      isFalse,
    );
    expect(
      canPublishAnnouncementMessage(
        remoteGuild,
        remoteSource,
        KaedeMessage.fromJson(<String, Object?>{
          ..._messageJson(),
          'origin_domain': 'remote.example',
          'channel_id': '6',
          'channel_domain': 'remote.example',
        }),
        actor,
      ),
      isTrue,
    );
  });

  test('repository uses the full follower and crosspost contract', () async {
    final follow = _followJson();
    final adapter = _AnnouncementAdapter(<_Reply>[
      _Reply(jsonEncode(<Object?>[follow])),
      _Reply(jsonEncode(follow)),
      const _Reply('{}'),
      _Reply(jsonEncode(_messageJson(flags: messageFlagCrossposted))),
    ]);
    final repository = KaedeRepository(
      KaedeApiClient(
        vault: const SessionVault(),
        httpClient: Dio()..httpClientAdapter = adapter,
      ),
    );
    final source = EntityRef.parse('2@chat.example');
    final target = EntityRef.parse('3@remote.example');
    final message = EntityRef.parse('10@chat.example');

    final listed = await repository.announcementFollowers(source);
    final created = await repository.followAnnouncement(source, target);
    await repository.deleteAnnouncementFollow(source, created.ref);
    final published = await repository.publishAnnouncement(source, message);

    expect(listed.single.targetChannel, target);
    expect(created.federated, isTrue);
    expect(isPublishedAnnouncement(published), isTrue);
    expect(
      adapter.requests.map((request) => request.method),
      <String>['GET', 'POST', 'DELETE', 'POST'],
    );
    expect(
      adapter.requests.map((request) => request.path),
      <String>[
        '/api/v1/channels/2@chat.example/followers',
        '/api/v1/channels/2@chat.example/followers',
        '/api/v1/channels/2@chat.example/followers/42%40remote.example',
        '/api/v1/channels/2@chat.example/messages/10@chat.example/crosspost',
      ],
    );
    expect(
      adapter.requests[1].data,
      <String, Object?>{'target_channel_id': '3@remote.example'},
    );
  });

  test('announcement responses reject substituted lineage and ambiguous state',
      () async {
    final source = EntityRef.parse('2@chat.example');
    final target = EntityRef.parse('3@remote.example');
    final valid = _followJson();
    for (final hostile in <Map<String, Object?>>[
      <String, Object?>{...valid, 'source_channel_id': '20'},
      <String, Object?>{...valid, 'target_channel_id': '30'},
      <String, Object?>{...valid, 'ref': '43@remote.example'},
      <String, Object?>{...valid, 'active': 'true'},
      <String, Object?>{...valid, 'updated_at': '2026-08-27T11:00:00Z'},
    ]) {
      expect(
        () => AnnouncementFollow.fromJson(
          hostile,
          expectedSource: source,
          expectedTarget: target,
        ),
        throwsFormatException,
      );
    }

    final adapter = _AnnouncementAdapter(<_Reply>[
      _Reply(jsonEncode(<Object?>[
        <String, Object?>{...valid, 'id': '43', 'ref': '43@remote.example'},
        valid,
      ])),
      _Reply(jsonEncode(<String, Object?>{
        ..._messageJson(flags: messageFlagCrossposted),
        'channel_id': '9',
      })),
      _Reply(jsonEncode(_messageJson(flags: 0))),
    ]);
    final repository = KaedeRepository(
      KaedeApiClient(
        vault: const SessionVault(),
        httpClient: Dio()..httpClientAdapter = adapter,
      ),
    );
    await expectLater(
        repository.announcementFollowers(source), throwsFormatException);
    await expectLater(
      repository.publishAnnouncement(
          source, EntityRef.parse('10@chat.example')),
      throwsFormatException,
    );
    await expectLater(
      repository.publishAnnouncement(
          source, EntityRef.parse('10@chat.example')),
      throwsFormatException,
    );
  });

  testWidgets('management UI explains missing source read permissions',
      (tester) async {
    final lockedGuild = _guild(
      '1',
      'Local',
      <KaedeChannel>[_channel(id: '2', type: 5, permissions: 0)],
    );
    final repository = KaedeRepository(
      KaedeApiClient(
        vault: const SessionVault(),
        httpClient: Dio()..httpClientAdapter = _AnnouncementAdapter(<_Reply>[]),
      ),
    );

    await tester.pumpWidget(MaterialApp(
      home: AnnouncementManagementTab(
        guild: lockedGuild,
        guilds: <KaedeGuild>[lockedGuild],
        currentUser: actor,
        repository: repository,
      ),
    ));

    expect(find.text('Followers are unavailable'), findsOneWidget);
    expect(
      find.textContaining('View Channel and Read Message History'),
      findsOneWidget,
    );
  });

  testWidgets('management UI exposes readable sources and eligible targets',
      (tester) async {
    final source = _channel(
      id: '2',
      type: 5,
      permissions: Permission.viewChannel | Permission.readMessageHistory,
    );
    final target = _channel(
      id: '3',
      type: 0,
      permissions: Permission.manageWebhooks,
    );
    final guild = _guild('1', 'Local', <KaedeChannel>[source, target]);
    final repository = KaedeRepository(
      KaedeApiClient(
        vault: const SessionVault(),
        httpClient: Dio()
          ..httpClientAdapter = _AnnouncementAdapter(
            <_Reply>[const _Reply('[]')],
          ),
      ),
    );

    await tester.pumpWidget(MaterialApp(
      home: AnnouncementManagementTab(
        guild: guild,
        guilds: <KaedeGuild>[guild],
        currentUser: actor,
        repository: repository,
      ),
    ));
    await tester.pumpAndSettle();

    expect(
      find.byKey(
        const ValueKey('announcement-source-picker-2@chat.example'),
      ),
      findsOneWidget,
    );
    expect(
      find.byKey(
        const ValueKey(
          'announcement-target-picker-2@chat.example-null',
        ),
      ),
      findsOneWidget,
    );
    expect(
      find.text('Destinations can be in another Kaede guild or instance.'),
      findsOneWidget,
    );
    expect(find.text('No follower channels yet'), findsOneWidget);
  });

  testWidgets('channel Follow surface fixes the source and hides management',
      (tester) async {
    final source = _channel(
      id: '2',
      type: 5,
      permissions: Permission.viewChannel | Permission.readMessageHistory,
    );
    final target = _channel(
      id: '3',
      type: 0,
      permissions: Permission.manageWebhooks,
    );
    final guild = _guild('1', 'Local', <KaedeChannel>[source, target]);
    final repository = KaedeRepository(
      KaedeApiClient(
        vault: const SessionVault(),
        httpClient: Dio()
          ..httpClientAdapter = _AnnouncementAdapter(
            <_Reply>[const _Reply('[]')],
          ),
      ),
    );

    await tester.pumpWidget(MaterialApp(
      home: AnnouncementManagementTab(
        guild: guild,
        guilds: <KaedeGuild>[guild],
        currentUser: actor,
        repository: repository,
        sourceChannel: source,
        createOnly: true,
      ),
    ));
    await tester.pumpAndSettle();

    expect(find.text('Follow #channel-2'), findsOneWidget);
    expect(
        find.byKey(const ValueKey('announcement-source-picker-2@chat.example')),
        findsNothing);
    expect(find.text('Follower channels'), findsNothing);
    expect(find.byKey(const ValueKey('announcement-follow-button')),
        findsOneWidget);
  });
}

KaedeChannel _channel({
  required String id,
  required int type,
  required int permissions,
  String domain = 'chat.example',
  String guildId = '1',
  bool e2eeRequired = false,
}) =>
    KaedeChannel.fromJson(<String, Object?>{
      'id': id,
      'origin_domain': domain,
      'guild_id': guildId,
      'guild_domain': domain,
      'type': type,
      'name': 'channel-$id',
      'position': int.parse(id),
      'permissions': '$permissions',
      'e2ee_required': e2eeRequired,
    });

KaedeGuild _guild(
  String id,
  String name,
  List<KaedeChannel> channels, {
  String domain = 'chat.example',
}) =>
    KaedeGuild.fromJson(<String, Object?>{
      'id': id,
      'origin_domain': domain,
      'name': name,
      'owner_id': '99',
      'owner_domain': domain,
      'permissions': '0',
      'unavailable': false,
      'channels': channels.map((channel) => channel.toJson()).toList(),
    });

Map<String, Object?> _followJson() => <String, Object?>{
      'id': '42',
      'ref': '42@remote.example',
      'source_channel_id': '2',
      'source_channel_domain': 'chat.example',
      'target_channel_id': '3',
      'target_channel_domain': 'remote.example',
      'creator_id': '7',
      'creator_domain': 'chat.example',
      'active': true,
      'federated': true,
      'generation': '1',
      'lifecycle_state': 'active',
      'name': null,
      'avatar_hash': null,
      'created_at': '2026-08-27T12:00:00Z',
      'updated_at': '2026-08-27T12:00:00Z',
    };

Map<String, Object?> _messageJson({int flags = 0}) => <String, Object?>{
      'id': '10',
      'origin_domain': 'chat.example',
      'channel_id': '2',
      'channel_domain': 'chat.example',
      'author_id': '7',
      'author_domain': 'chat.example',
      'author': <String, Object?>{
        'id': '7',
        'origin_domain': 'chat.example',
        'username': 'maple',
        'handle': 'maple@chat.example',
      },
      'content': 'News',
      'message_type': 0,
      'flags': flags,
      'created_at': '2026-08-27T12:00:00Z',
    };

final class _Reply {
  const _Reply(this.body);

  final String body;
}

final class _AnnouncementAdapter implements HttpClientAdapter {
  _AnnouncementAdapter(this._replies);

  final List<_Reply> _replies;
  final List<RequestOptions> requests = <RequestOptions>[];

  @override
  Future<ResponseBody> fetch(
    RequestOptions options,
    Stream<Uint8List>? requestStream,
    Future<void>? cancelFuture,
  ) async {
    requests.add(options);
    final reply = _replies.removeAt(0);
    return ResponseBody.fromString(
      reply.body,
      200,
      headers: <String, List<String>>{
        Headers.contentTypeHeader: <String>[Headers.jsonContentType],
      },
    );
  }

  @override
  void close({bool force = false}) {}
}

File: frontend/src/lib/chat/announcements.test.ts
import { beforeEach, describe, expect, it, vi } from 'vitest';

const apiMock = vi.hoisted(() => vi.fn());
vi.mock('$lib/api/client', () => ({ api: apiMock }));

import { Permission } from '$lib/generated/permissions';
import {
  announcementTargets,
  canReadAnnouncementChannel,
  canPublishAnnouncementMessage,
  channelFollowSystemMessageText,
  createAnnouncementFollow,
  deleteAnnouncementFollow,
  isPublishedAnnouncement,
  listAnnouncementFollows,
  MESSAGE_FLAG_CROSSPOSTED,
  publishAnnouncementMessage
} from './announcements';
import type { Channel, Guild, Message, UserSummary } from './types';

const user: UserSummary = {
  id: '7',
  origin_domain: 'chat.example',
  username: 'maple',
  display_name: null,
  avatar_hash: null,
  handle: 'maple@chat.example'
};

function channel(id: string, type: number, permissions: bigint, domain = 'chat.example'): Channel {
  return {
    id,
    origin_domain: domain,
    guild_id: '1',
    guild_domain: domain,
    type,
    name: `channel-${id}`,
    topic: null,
    position: Number(id),
    parent_id: null,
    parent_domain: null,
    permissions: permissions.toString(),
    rate_limit_per_user: 0,
    last_message_id: null,
    last_message_domain: null
  };
}

function guild(id: string, name: string, channels: Channel[]): Guild {
  return {
    id,
    origin_domain: channels[0]?.origin_domain ?? 'chat.example',
    name,
    description: null,
    icon_hash: null,
    owner_id: '9',
    permission_generation: '1',
    unavailable: false,
    channels
  };
}

function message(overrides: Partial<Message> = {}): Message {
  return {
    id: '10',
    origin_domain: 'chat.example',
    channel_id: '2',
    channel_domain: 'chat.example',
    author_id: user.id,
    author_domain: user.origin_domain,
    author: user,
    content: 'News',
    message_type: 0,
    flags: 0,
    client_nonce: null,
    referenced_message_id: null,
    referenced_message_domain: null,
    mention_user_refs: [],
    edited_at: null,
    deleted_at: null,
    created_at: '2026-08-27T12:00:00Z',
    ...overrides
  };
}

function follow() {
  return {
    id: '42',
    ref: '42@remote.example',
    source_channel_id: '2',
    source_channel_domain: 'chat.example',
    target_channel_id: '3',
    target_channel_domain: 'remote.example',
    creator_id: '7',
    creator_domain: 'chat.example',
    active: true,
    federated: true,
    generation: '1',
    lifecycle_state: 'active',
    name: null,
    avatar_hash: null,
    created_at: '2026-08-27T12:00:00Z',
    updated_at: '2026-08-27T12:00:00Z'
  };
}

describe('announcement client', () => {
  beforeEach(() => apiMock.mockReset());

  it('allows announcement-follow discovery with View Channel alone', () => {
    const news = channel('2', 5, Permission.VIEW_CHANNEL);

    expect(canReadAnnouncementChannel(news, guild('1', 'Local', [news]))).toBe(true);
    expect(
      canReadAnnouncementChannel(
        channel('3', 5, Permission.READ_MESSAGE_HISTORY),
        guild('1', 'Local', [news])
      )
    ).toBe(false);
  });

  it('uses canonical encoded routes for list, create, delete, and publish', async () => {
    apiMock
      .mockResolvedValueOnce([follow()])
      .mockResolvedValueOnce(follow())
      .mockResolvedValueOnce(undefined)
      .mockResolvedValueOnce({ ...message(), flags: MESSAGE_FLAG_CROSSPOSTED });
    const source = channel('2', 5, Permission.ADMINISTRATOR);
    const post = message();
    await listAnnouncementFollows('2@chat.example');
    await createAnnouncementFollow('2@chat.example', '3@remote.example');
    await deleteAnnouncementFollow('2@chat.example', '42@remote.example');
    await publishAnnouncementMessage(source, post);

    expect(apiMock.mock.calls).toEqual([
      ['/channels/2%40chat.example/followers', { signal: undefined }],
      [
        '/channels/2%40chat.example/followers',
        { method: 'POST', body: JSON.stringify({ target_channel_id: '3@remote.example' }) }
      ],
      ['/channels/2%40chat.example/followers/42%40remote.example', { method: 'DELETE' }],
      ['/channels/2%40chat.example/messages/10%40chat.example/crosspost', { method: 'POST' }]
    ]);
  });

  it('rejects substituted follower and publish response lineage', async () => {
    apiMock.mockResolvedValueOnce({ ...follow(), target_channel_id: '30' });
    await expect(createAnnouncementFollow('2@chat.example', '3@remote.example')).rejects.toThrow(
      /requested lineage/u
    );

    apiMock.mockResolvedValueOnce([{ ...follow(), id: '43', ref: '43@remote.example' }, follow()]);
    await expect(listAnnouncementFollows('2@chat.example')).rejects.toThrow(/unordered/u);

    apiMock.mockResolvedValueOnce({ ...message(), channel_id: '20' });
    await expect(
      publishAnnouncementMessage(channel('2', 5, Permission.ADMINISTRATOR), message())
    ).rejects.toThrow(/requested lineage/u);

    apiMock.mockResolvedValueOnce({ ...message(), flags: 0 });
    await expect(
      publishAnnouncementMessage(channel('2', 5, Permission.ADMINISTRATOR), message())
    ).rejects.toThrow(/requested lineage/u);
  });

  it('offers only plaintext regular text targets with Manage Webhooks', () => {
    const manage = Permission.MANAGE_WEBHOOKS;
    const allowedText = channel('3', 0, manage);
    const remoteText = channel('4', 0, Permission.ADMINISTRATOR, 'remote.example');
    const deniedNews = channel('8', 5, Permission.ADMINISTRATOR, 'remote.example');
    const denied = channel('5', 0, Permission.VIEW_CHANNEL);
    const voice = channel('6', 2, manage);
    const encrypted = { ...channel('7', 0, manage), encryption_mode: 'e2ee' as const };
    const e2eeRequired = { ...channel('9', 0, manage), e2ee_required: true };

    expect(
      announcementTargets([
        guild('1', 'Local', [denied, allowedText, voice, encrypted, e2eeRequired]),
        guild('2', 'Remote', [remoteText, deniedNews])
      ]).map((target) => target.ref)
    ).toEqual(['3@chat.example', '4@remote.example']);
  });

  it('matches publish ownership, moderation, authority, encryption, and flag rules', () => {
    const source = channel('2', 5, Permission.SEND_MESSAGES);
    expect(canPublishAnnouncementMessage(source, message(), user, true, false)).toBe(true);
    expect(
      canPublishAnnouncementMessage(source, message({ author_id: '8' }), user, true, false)
    ).toBe(false);
    expect(
      canPublishAnnouncementMessage(source, message({ author_id: '8' }), user, true, true)
    ).toBe(true);
    expect(
      canPublishAnnouncementMessage(
        { ...source, origin_domain: 'remote.example' },
        message({ channel_domain: 'remote.example' }),
        user,
        true,
        true
      )
    ).toBe(true);
    expect(canPublishAnnouncementMessage(source, message({ flags: 1 }), user, true, true)).toBe(
      false
    );
    expect(canPublishAnnouncementMessage(source, message({ e2ee: {} }), user, true, true)).toBe(
      false
    );
    expect(isPublishedAnnouncement(message({ flags: 3 }))).toBe(true);
  });

  it('renders Discord type-12 follow notices from known or federated channel references', () => {
    const follow = message({
      content: 'upstream-news',
      message_type: 12,
      message_reference: {
        type: 0,
        channel_id: '8',
        channel_domain: 'remote.example',
        guild_id: '9',
        guild_domain: 'remote.example'
      }
    });
    const known = channel('8', 5, Permission.VIEW_CHANNEL, 'remote.example');
    known.name = 'release-notes';

    expect(channelFollowSystemMessageText(follow, [known], 'Maple')).toBe(
      'Maple has added #release-notes to this channel. Its most important updates will show up here.'
    );
    expect(channelFollowSystemMessageText(follow, [], 'Maple')).toContain('#upstream-news');
    expect(
      channelFollowSystemMessageText(message({ ...follow, content: null }), [], 'Maple')
    ).toContain('#8@remote.example');
    expect(channelFollowSystemMessageText(message(), [], 'Maple')).toBeNull();
  });
});
